Mathlib/Topology/TietzeExtension.lean

```diff
@@ -0,0 +1,435 @@
+a bounded continuous real-valued function `g : Y →ᵇ ℝ` such that `g y ∈ t` for all `y` and
+`g ∘ e = f`. -/
+theorem exists_extension_forall_mem_of_closedEmbedding (f : X →ᵇ ℝ) {t : Set ℝ} {e : X → Y}
+    [hs : OrdConnected t] (hf : ∀ x, f x ∈ t) (hne : t.Nonempty) (he : ClosedEmbedding e) :
+    ∃ g : Y →ᵇ ℝ, (∀ y, g y ∈ t) ∧ g ∘ e = f := by
+  cases isEmpty_or_nonempty X
+  · rcases hne with ⟨c, hc⟩
+    refine' ⟨const Y c, fun _ => hc, funext fun x => isEmptyElim x⟩
+  rcases exists_extension_forall_exists_le_ge_of_closedEmbedding f he with ⟨g, hg, hgf⟩
+  refine' ⟨g, fun y => _, hgf⟩
+  rcases hg y with ⟨xl, xu, h⟩
+  exact hs.out (hf _) (hf _) h
+#align bounded_continuous_function.exists_extension_forall_mem_of_closed_embedding BoundedContinuousFunction.exists_extension_forall_mem_of_closedEmbedding
+
+/-- **Tietze extension theorem** for real-valued bounded continuous maps, a version for a closed
+set. Let `s` be a closed set in a normal topological space `Y`. Let `f` be a bounded continuous
+real-valued function on `s`. Let `t` be a nonempty convex set of real numbers (we use
+`ord_connected` instead of `convex` to automatically deduce this argument by typeclass search) such
+that `f x ∈ t` for all `x : s`. Then there exists a bounded continuous real-valued function
+`g : Y →ᵇ ℝ` such that `g y ∈ t` for all `y` and `g.restrict s = f`. -/
+theorem exists_forall_mem_restrict_eq_of_closed {s : Set Y} (f : s →ᵇ ℝ) (hs : IsClosed s)
+    {t : Set ℝ} [OrdConnected t] (hf : ∀ x, f x ∈ t) (hne : t.Nonempty) :
+    ∃ g : Y →ᵇ ℝ, (∀ y, g y ∈ t) ∧ g.restrict s = f := by
+  rcases exists_extension_forall_mem_of_closedEmbedding f hf hne
+      (closedEmbedding_subtype_val hs) with
+    ⟨g, hg, hgf⟩
+  exact ⟨g, hg, FunLike.coe_injective hgf⟩
+#align bounded_continuous_function.exists_forall_mem_restrict_eq_of_closed BoundedContinuousFunction.exists_forall_mem_restrict_eq_of_closed
+
+end BoundedContinuousFunction
+
+namespace ContinuousMap
+
+set_option maxHeartbeats 0 in
+/-- **Tietze extension theorem** for real-valued continuous maps, a version for a closed
+embedding. Let `e` be a closed embedding of a nonempty topological space `X` into a normal
+topological space `Y`. Let `f` be a continuous real-valued function on `X`. Let `t` be a nonempty
+convex set of real numbers (we use `ord_connected` instead of `convex` to automatically deduce this
+argument by typeclass search) such that `f x ∈ t` for all `x`. Then there exists a continuous
+real-valued function `g : C(Y, ℝ)` such that `g y ∈ t` for all `y` and `g ∘ e = f`. -/
+theorem exists_extension_forall_mem_of_closedEmbedding (f : C(X, ℝ)) {t : Set ℝ} {e : X → Y}
+    [hs : OrdConnected t] (hf : ∀ x, f x ∈ t) (hne : t.Nonempty) (he : ClosedEmbedding e) :
+    ∃ g : C(Y, ℝ), (∀ y, g y ∈ t) ∧ g ∘ e = f := by
+  have h : ℝ ≃o Ioo (-1 : ℝ) 1 := orderIsoIooNegOneOne ℝ
+  set F : X →ᵇ ℝ :=
+    { toFun := Subtype.val ∘ h ∘ f
```

Comment:
This was originally `coe ∘ h ∘ f`. I'm guessing that this definition is causing the `masHeartbeats` error.

Reply:
I'm not able to find the direct analogue of this `coe` in `mathlib4`.

Reply:
This is the output of `set_option profiler true`

```lean 
tactic execution of Lean.Parser.Tactic.rewriteSeq took 164ms
typeclass inference of CoeT took 282ms
typeclass inference of CoeT took 277ms
typeclass inference of CoeT took 277ms
typeclass inference of CoeT took 283ms
typeclass inference of CoeT took 282ms
typeclass inference of CoeT took 276ms
typeclass inference of CoeT took 275ms
typeclass inference of CoeT took 281ms
typeclass inference of CoeT took 289ms
typeclass inference of CoeT took 281ms
typeclass inference of CoeT took 285ms
typeclass inference of CoeT took 273ms
typeclass inference of CoeT took 275ms
typeclass inference of CoeT took 277ms
typeclass inference of CoeT took 278ms
typeclass inference of CoeT took 280ms
typeclass inference of CoeT took 276ms
typeclass inference of CoeT took 275ms
typeclass inference of CoeT took 278ms
typeclass inference of CoeT took 277ms
typeclass inference of CoeT took 278ms
typeclass inference of CoeT took 276ms
typeclass inference of CoeT took 274ms
typeclass inference of CoeT took 273ms
typeclass inference of CoeT took 274ms
tactic execution of Lean.Parser.Tactic.rewriteSeq took 230ms
```

Reply:
The true cause seems in `set` tactics.
This command replace declared subterms in local hypotheses and goal with the new definition.
Actually, all `set` tactics in this theorem can be replaced with `let`, and it solves timeouts.